### PR TITLE
cli: add `--disable-max-offset-check` flag

### PR DESCRIPTION
**Pull request by erikgrinaker**
**util/hlc: add `Clock.ToleratedOffset()`**

This patch adds `Clock.ToleratedOffset()` which returns the tolerated clock offset as measured via RPC heartbeats before the node will self-terminate. The existing hard-coded 80% tolerated clock skew is moved from `RemoteOffset.isHealth()` to `hlc.NewClock()` and passed in via `rpc.ContextOptions.MaxOffset`.

As a consequence of setting `rpc.ContextOptions.MaxOffset` to `ToleratedOffset()` rather than `MaxOffset()`, we now discard RPC clock measurements when the RPC heartbeat round-trip latency exceeds 1.6x of the max offset (2x the tolerated offset), rather than 2x of the max offset. A later commit will rename the `MaxOffset` option to `ToleratedOffset`.

Release note: None
  
**rpc: rename `MaxOffset` to `ToleratedOffset`**

Mechanical rename, except for changes to a couple of log/error messages.

Release note: None
  
**rpc: explicitly allow `ToleratedOffset = 0`**

This patch explicitly allows `ToleratedOffset = 0`, which disables RPC clock offset checking and is often done in tests. The relevant code is also unindented. There are no behavioral changes.

Release note: None
  
**hlc: add and use `NewClockForTesting()`**

This patch adds `NewClockForTesting()`, which returns a clock with no max offset or tolerated offset. It takes a wall clock as a parameter, defaulting to the system clock if `nil`.

All appropriate tests have been updated to use it.

Release note: None

**hlc: add `toleratedOffset` parameter for `NewClock()`**

The tolerated offset is computed via `BaseConfig.ToleratedOffset()`, preserving the previous value of 80% of `MaxOffset`.

Epic: none
Release note: None
  
**cli: add `--disable-max-offset-check` flag**

This patch adds a `--disable-max-offset-check` flag, which disables the clock offset check that will self-terminate the node if its offset to the rest of the cluster (as measured via RPC heartbeats) exceeds `--max-offset`.

This is motivated by deployments with reliable, high-precision clock infrastructure where `--max-offset` may be set very low (e.g. 10 ms) to reduce undertainty restarts and write latency for global tables. With such low offset limits, RPC latency spikes could cause spurious node restarts. This flag allows operators to assume responsibility for ensuring real clock skew never exceeds `--max-offset`.

Resolves #94999.
Epic: none

Release note (ops change): The flag `--disable-max-offset-check` has been added to disable node self-termination when it detects clock skew with the rest of the cluster beyond `--max-offset`. The operator assumes responsibility for ensuring that real clock skew never exceeds `--max-offset`.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
> Reviewed 4 of 19 files at r3, 2 of 77 files at r5, 3 of 14 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96141)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @JeffSwenson, @nvanbenschoten, and @tbg)

___
*[`pkg/server/config.go` line 270 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/96141#-NMtFAG_AnloWgzC9AtE:-NMtFAG_AnloWgzC9AtF:b-2fy88r) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a7509e40d2acb87b0d445e2a694716af283222/pkg/server/config.go#L270)):*
> ```Go
> 	cfg.ClusterIDContainer = idsProvider.clusterID
> 	cfg.AmbientCtx = log.MakeServerAmbientContext(tr, idsProvider)
> 	cfg.MaxOffset = OffsetType(base.DefaultMaxClockOffset)
> ```

You need to assign a default value to the new field here, even if it's just zero. This reset function is used in-between CLI tests to restore the initial state.

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96141)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, @nvanbenschoten, and @tbg)

___
*[`pkg/server/config.go` line 270 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/96141#-NMtFAG_AnloWgzC9AtE:-NN0xmho-zfnN-HVGxfP:b-yvvs31) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a7509e40d2acb87b0d445e2a694716af283222/pkg/server/config.go#L270)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

You need to assign a default value to the new field here, even if it's just zero. This reset function is used in-between CLI tests to restore the initial state.
</blockquote></details>

Good call, thanks -- done.

**Review by tbg — APPROVED** [review]
The code LGTM, just have a basic question.

Is it true that with a very small tolerated clock offset, the RPC latencies can lead to erroneous node fatal errors, which motivates this work? I thought we recorded the "uncertainty" in these measurements to avoid precisely that:

https://github.com/cockroachdb/cockroach/blob/c85ca9be8bc8d46d56680864dc0503979178d759/pkg/rpc/context.go#L2352

I would think that instead of causing extra failures, the latencies would *mask* failures, wouldn't they? So do we know we actually need the `--tolerated-offset` flag?

Even if we don't I think this PR is useful, but I want to understand the requirements.

I pushed an additional commit with comment suggestions that I wrote while doing some sleuthing in this PR to clear things up for myself. If you don't find them helpful give me feedback and I'll work them over. You're also welcome to patch them up yourself, of course.

File: pkg/util/hlc/hlc.go [hunk 1]
```diff
@@ -52,9 +52,28 @@ type Clock struct {
 	// RPC heartbeats compare detected clock skews against this value to protect
```
[the remark this answers is not in the document]
Comment (tbg) [reply pkg/util/hlc/hlc.go]: I took a pass at this myself. And realized the "invariant" is never true today ;-)

**Review by tbg — APPROVED** [review]
Forgot a commit, still LGTM

**Comment by erikgrinaker** [comment]
> Is it true that with a very small tolerated clock offset, the RPC latencies can lead to erroneous node fatal errors, which motivates this work? I thought we recorded the "uncertainty" in these measurements to avoid precisely that:
> 
> https://github.com/cockroachdb/cockroach/blob/c85ca9be8bc8d46d56680864dc0503979178d759/pkg/rpc/context.go#L2352
> 
> I would think that instead of causing extra failures, the latencies would _mask_ failures, wouldn't they? So do we know we actually need the `--tolerated-offset` flag?

I think this assumes that network latency is symmetric for the outbound and inbound packets. So if e.g. the outbound packet saw packet loss and was delayed by 200 ms for the retransmit, but the return packet was sent in 1 ms, then the uncertainty would be 100 ms (201 ms / 2), while the remote clock reading was taken 200 ms after the ping was sent.

I haven't looked into the exact consequences of this, and it's probably worthwhile doing so, but I couldn't confidently say that dropping the max offset to 10 ms wouldn't cause any spurious restarts. I'll have a closer look when I get a chance.

**Comment by erikgrinaker** [comment]
I've updated this PR to implement `--disable-max-offset-check` (disabling RPC offset checks) rather than `--tolerate-offset` (which allowed a custom tolerated offset). The tolerated offset is now hardcoded to 80% of `--max-offset` like before, unless disabled.

PTAL.

File: pkg/server/config.go [hunk 2]
```diff
@@ -89,30 +89,30 @@ var productionSettingsWebpage = fmt.Sprintf(
...
-func (mo *MaxOffsetType) Set(v string) error {
+func (o *OffsetType) Set(v string) error {
 	nanos, err := time.ParseDuration(v)
 	if err != nil {
 		return err
 	}
-	if nanos > maximumMaxClockOffset {
-		return errors.Errorf("%s is not a valid max offset, must be less than %v.", v, maximumMaxClockOffset)
+	if nanos > maximumClockOffset {
+		return errors.Errorf("%s is not a valid offset, must be less than %v.", v, maximumClockOffset)
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pkg/server/config.go]: We lowered the duration to 6 seconds, so we try to extend them 3 seconds in. We probably shouldn't allow setting it to much more than 2 seconds?

**Review by tbg — APPROVED** [review]

**Pull request merged**